lambda/matching.ml

```diff
@@ -944,31 +992,48 @@ end = struct
     if Context.is_empty ctx then
       jumps
     else
-      add jumps
-
-  let rec union (env1 : t) env2 =
-    match (env1, env2) with
-    | [], _ -> env2
-    | _, [] -> env1
-    | ((i1, pss1) as x1) :: rem1, ((i2, pss2) as x2) :: rem2 ->
-        if i1 = i2 then
-          (i1, Context.union pss1 pss2) :: union rem1 rem2
-        else if i1 > i2 then
-          x1 :: union rem1 env2
-        else
-          x2 :: union env1 rem2
+      { jumps with env = add jumps.env }
+
+  let singleton i ctx =
+    (* Total: a singleton only jumps to exit [i],
+       not to the final exit. *)
+    add i ctx (empty Total)
```

Comment:
This (and the previous issue with final exit ending up bound in the jump summary through `Jumps.add`) suggests an awkward corner of our API: one should not call `Jumps.singleton` or `Jumps.add` with the final exit, but there is no easy way to check this inside the `Jumps` module because jump summaries do not know what the final exit number is.

I agree that this is awkward and worth fixing, and I have two fixes in mind:

1. I could make the jump-summary-extension functions take the default environment as parameter, to check dynamically that the exit is a non-final exit of the default environment, or
2. I could change the API of `Default_environment` to *hide* the final exit better. The only way it is used during compilation is to produce a fail action, so I could just add a `Default_environment.final_fail_action` function and that's it. (This would require remodeling `mk_failaction_pos` a bit, possibly for the better.) Then it would be very obvious that the exit number cannot possibly be the final exit, as there would be no way to extract it from the default environment.

I prefer option (2) and will give it a try after mulling over it.

Reply:
(1) seems simple enough too. The dynamic check could just be "raise if they're the same". I don't want to indicate that I prefer a significant API overhaul when the API is just used in a few places, and is used correctly in all of those places.

Reply:
I pushed a new commit that performs a refactoring along (2). Let me know what you think. If we decide to go in this direction, I will rebase the PR to squash it into the first commit.

Reply:
I think I prefer your original approach. It's easier for me to judge its correctness, and I don't find the API change to make substantially more-useful guarantees.

I may just suggest a comment along the lines of "the label must not be the final exit" on both `Jumps.singleton`/`Jumps.add`, but even that I don't feel strongly about.

Reply:
I worked on `mk_failaction_pos` again but in the end I decided to keep the approach with `raise_final_exit` (so not the original approach), I find it cleaner -- and it is not more complex.

Reply:
OK, sounds good. Thanks.